### PR TITLE
Azure deployment glue

### PR DESCRIPTION
**Pull request by asyasky**
Adds/updates the things we need to do an Azure Deployment using Azure DevOps Continuous Deployment. Sets up the framework for using Storage & KeyVault.

### CODE REVIEW

File: ServerCore/Properties/PublishProfiles/PuzzleServerTestDeploy - Web Deploy.pubxml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (morganbr) [inline ServerCore/Properties/PublishProfiles/PuzzleServerTestDeploy - Web Deploy.pubxml]: You seem to have two of these. Do we actually need both?

File: ServerCore/Properties/launchSettings.json [hunk 2]
```diff
@@ -12,16 +12,20 @@
...
+        "ASPNETCORE_HOSTINGSTARTUP__KEYVAULT__CONFIGURATIONENABLED": "true",
+        "ASPNETCORE_HOSTINGSTARTUP__KEYVAULT__CONFIGURATIONVAULT": "https://PuzzleServerTestKeyVault.vault.azure.net"
       }
     },
     "ServerCore": {
       "commandName": "Project",
       "launchBrowser": true,
       "environmentVariables": {
-        "ASPNETCORE_ENVIRONMENT": "Development"
+        "ASPNETCORE_ENVIRONMENT": "Development",
```
Comment (morganbr) [inline ServerCore/Properties/launchSettings.json]: Do you need to leave this on for now? If so, please file an issue to turn it off

File: ServerCore/Startup.cs [hunk 3]
```diff
@@ -14,6 +15,17 @@ public Startup(IConfiguration configuration)
             Configuration = configuration;
         }
 
+
```
Comment (morganbr) [inline ServerCore/Startup.cs]: Extra white space

File: ServerCore/Startup.cs [hunk 4]
```diff
@@ -28,28 +40,41 @@ public void ConfigureServices(IServiceCollection services)
                     options.Conventions.AuthorizeFolder("/Teams");
                 });
 
-            services.AddDbContext<PuzzleServerContext>
-                (options => options.UseLazyLoadingProxies()
-                    .UseSqlServer(Configuration.GetConnectionString("PuzzleServerContext")));
+            // Use SQL Database if in Azure, otherwise, use localdb
```
Comment (morganbr) [inline ServerCore/Startup.cs]: This looks like duplication of Program.cs. Which one do we actually need?

File: ServerCore/Startup.cs [hunk 5]
```diff
@@ -14,6 +15,17 @@ public Startup(IConfiguration configuration)
             Configuration = configuration;
         }
 
+
+        public Startup(IHostingEnvironment env)
```
Comment (morganbr) [inline ServerCore/Startup.cs]: Also looks like a dupe of Program.cs

File: ServerCore/Startup.cs [hunk 6]
```diff
@@ -28,28 +40,41 @@ public void ConfigureServices(IServiceCollection services)
...
+
+            if (env.IsDevelopment() || Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") == "Development")
             {
                 app.UseBrowserLink();
                 app.UseDeveloperExceptionPage();
-                PuzzleServerContext.UpdateDatabase(app);
             }
             else
             {
+                app.UseDeveloperExceptionPage();
```
Comment (morganbr) [inline ServerCore/Startup.cs]: Please remove this or file an issue to do so

File: ServerCore/Startup.cs [hunk 7]
```diff
@@ -28,28 +40,41 @@ public void ConfigureServices(IServiceCollection services)
...
+                microsoftOptions.ClientSecret = Configuration["Authentication-Microsoft-Password"];
+
             });
         }
 
         // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
         public void Configure(IApplicationBuilder app, IHostingEnvironment env)
         {
-            if (env.IsDevelopment())
+            PuzzleServerContext.UpdateDatabase(app);
```
Comment (morganbr) [inline ServerCore/Startup.cs]: Is this how we actually want production db updates to happen?

File: ServerCore/Areas/Identity/IdentityHostingStartup.cs [hunk 8]
```diff
@@ -14,9 +15,28 @@ public void Configure(IWebHostBuilder builder)
...
+                    IConfigurationBuilder configBuilder = new ConfigurationBuilder()
+                        .SetBasePath(env.ContentRootPath)
+                        .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
+                        .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
+                        .AddEnvironmentVariables();
+                    context.Configuration = configBuilder.Build();
+
+                    services.AddDbContext<PuzzleServerContext>
+                        (options => options.UseLazyLoadingProxies()
+                            .UseSqlServer(context.Configuration.GetConnectionString("PuzzleServerSQLConnectionString")));
```
Comment (morganbr) [inline ServerCore/Areas/Identity/IdentityHostingStartup.cs]: Please document where our secrets are and how to set them up as part of this so we can set up real production if we need it

File: ServerCore/Properties/launchSettings.json (hunk 2 above)
Comment (asyasky) [reply ServerCore/Properties/launchSettings.json]: This is the only way to get useful error messages right now. I'm looking into other ways for the long term (supposedly I can turn on logging in the web.config, but the web.config is generated during Publish & I haven't traced that back to what I actually need to do). I'll add an issue.

File: ServerCore/Startup.cs (hunk 7 above)
Comment (asyasky) [reply ServerCore/Startup.cs]: That is an excellent question and I don't have an answer right now. This is the simplest way to take care of the database, but it does prevent us from having full control once we have real data.

File: ServerCore/Areas/Identity/IdentityHostingStartup.cs (hunk 8 above)
Comment (asyasky) [reply ServerCore/Areas/Identity/IdentityHostingStartup.cs]: The secrets store isn't set up yet, right now we're using the Azure Application Settings - a future change will turn on KeyVault access and I'll add docs for using that.

File: ServerCore/Properties/PublishProfiles/PuzzleServerTestDeploy - Web Deploy.pubxml (hunk 1 above)
Comment (asyasky) [reply ServerCore/Properties/PublishProfiles/PuzzleServerTestDeploy - Web Deploy.pubxml]: Nope, good catch, thanks! :)

File: ServerCore/Startup.cs (hunk 6 above)
Comment (asyasky) [reply ServerCore/Startup.cs]: Removed (was for debugging)

**Pull request merged**